### PR TITLE
prov/sockets: Add support for IPv6 addressing

### PR DESCRIPTION
**Pull request by shefty**
Changes do not work.
Posting as work in progress to add ipv6 support to the socket provider. These changes currently break MSG EP support and have not been tested with ipv6 addresses yet.

### CODE REVIEW

File: prov/sockets/src/sock_ep_msg.c [hunk 1]
```diff
@@ -255,112 +255,84 @@ static int sock_ep_cm_getname(fid_t fid, void *addr, size_t *addrlen)
...
-		addr_size = sizeof(addr);
-		if (getsockname(pep->cm.sock, (struct sockaddr *)&addr, &addr_size))
-			return -FI_EINVAL;
-		pep->src_addr.sin_port = addr.sin_port;
-		sprintf(sa_port, "%d", ntohs(pep->src_addr.sin_port));
+		ret = -ofi_sockerr();
+		goto err;
 	}
 
-	if (pep->src_addr.sin_addr.s_addr == 0) {
```
Comment (shefty) [inline prov/sockets/src/sock_ep_msg.c]: Memo to me: This is setting a src_addr if it is a wildcard address.  Why?

**Comment by dmitrygladkov** [comment]
@shefty I'm testing IPv6 support. It seems that this almost works. I see that IPv6 is correctly reported, but there is some bug. I'll notify when I managed to find out a solution
image
ofi_sizeofaddr() reports "Unknown address"
But fi_getname works fine

**Comment by shefty** [comment]
I wasn't able to pass the fabtests MSG EP tests with my latest changes.  I haven't tried yet with your fix that I just merged.

**Comment by dmitrygladkov** [comment]
It should work now I guess.
Btw, Travis complains about a compilation error

**Comment by dmitrygladkov** [comment]
@shefty Many thanks! I tested this in my environment. It works as expected. Do you have some problems with fabtests over IPv6?

**Comment by shefty** [comment]
The shared context test does not pass.  All others are passing.  I haven't narrowed down the shared context failure yet (distracted yesterday), but what I've seen is that the contexts are picking up ipv4 addresses, which confuses the AV with a mix of ipv4 and ipv6.  The socket provider's handling of addresses through fi_getinfo is (or was) convoluted.  :)

And it seems that I have an issue with a debug print message that's causing travis to fail.  But that may be in a throw-away change that I'm using to debug.  Let me fix that, and if everything else passes, I'll look at merging what I have and fix the shared context problem in a separate PR.

**Comment by dmitrygladkov** [comment]
> The socket provider's handling of addresses through fi_getinfo is (or was) convoluted. :)

Yes, this's true :)


> I'll look at merging what I have and fix the shared context problem in a separate PR.

Sure. Thanks!

Btw, there is some issue with macOS testing in Travis. Did some tests hang? If needed I can test/debug on our MAC machines.

**Comment by dmitrygladkov** [comment]
> If you can debug the MAC failures, I would appreciate it.

Sure.

**Comment by shefty** [comment]
Testing locally, this allows the socket provider to work with ipv6 addressing.  The last fabtests that was failing was a problem in the test.

We need to figure out the OS X failure, then we can merge.

**Comment by dmitrygladkov** [comment]
> We need to figure out the OS X failure, then we can merge.

Sean, sorry for the delay with investigating OS X failure. I'll inform when it will be done -- probably, today or tomorrow

**Comment by dmitrygladkov** [comment]
@shefty just do not duplicate our work -- I've started analyzing the failures on OS X machines.

**Pull request merged**